packages/accounts-controller/src/AccountsController.ts

```diff
@@ -717,41 +701,57 @@ export class AccountsController extends BaseController<
         }
       }
 
-      if (deletedAccounts.length > 0) {
-        for (const account of deletedAccounts) {
-          this.#handleAccountRemoved(account.id);
+      this.update((currentState: Draft<AccountsControllerState>) => {
+        if (deletedAccounts.length > 0) {
+          for (const account of deletedAccounts) {
+            currentState.internalAccounts.accounts = this.#handleAccountRemoved(
+              currentState.internalAccounts.accounts,
+              account.id,
+            );
+          }
         }
-      }
 
-      if (addedAccounts.length > 0) {
-        for (const account of addedAccounts) {
-          this.#handleNewAccountAdded(account);
+        if (addedAccounts.length > 0) {
+          for (const account of addedAccounts) {
+            currentState.internalAccounts.accounts =
+              this.#handleNewAccountAdded(
+                currentState.internalAccounts.accounts,
+                account,
+              );
+          }
         }
-      }
 
-      // handle if the selected account was deleted
-      if (!this.getAccount(this.state.internalAccounts.selectedAccount)) {
-        const [accountToSelect] = this.listAccounts().sort(
-          (accountA, accountB) => {
-            // sort by lastSelected descending
-            return (
-              (accountB.metadata.lastSelected ?? 0) -
-              (accountA.metadata.lastSelected ?? 0)
-            );
-          },
+        // We don't use list accounts because it is not the updated state yet.
+        const existingAccounts = Object.values(
+          currentState.internalAccounts.accounts,
         );
 
-        // if the accountToSelect is undefined, then there are no accounts
-        // it mean the keyring was reinitialized.
-        if (!accountToSelect) {
-          this.update((currentState: Draft<AccountsControllerState>) => {
+        // handle if the selected account was deleted
+        if (
+          !currentState.internalAccounts.accounts[
+            this.state.internalAccounts.selectedAccount
+          ]
+        ) {
+          // if the accountToSelect is undefined, then there are no accounts
+          // it mean the keyring was reinitialized.
+          if (existingAccounts.length === 0) {
```

Comment:
There is no more `accountToSelect` here:

```suggestion
          // if currently selected account is undefined and there are no accounts
          // it mean the keyring was reinitialized.
          if (existingAccounts.length === 0) {
```